CDS-58185-https-polling (#1784)
harness/developer-hub: docs/continuous-delivery/x-platform-cd-features/cd-steps/utilities/http-step.md

> This results in a `401 Unauthorized` response due to an incorrect api key. However, the capability check will be successful and the task will be assigned to the Harness delegate. 

Reviewer: This is outside the scope of this review, but I think we need a correction here. Please convert "api" to all caps. 
If you have a few minutes, it might be good to check elsewhere or maybe add this to your bug bash as something to look for across your docs.

> This results in a `401 Unauthorized` response due to an incorrect api key. However, the capability check will be successful and the task will be assigned to the Harness delegate. 

Author: ```suggestion
This results in a `401 Unauthorized` response due to an incorrect API key. However, the capability check will be successful and the task will be assigned to the Harness delegate. 
```

Capability checks are basic accessibility checks and do not follow multiple redirects. Hence, Harness returns from the first `302 Found` response during capability checks. 

## HTTP polling

> HTTP step supports polling. When you create the HTTP step for polling, the client will keep requesting the resource at regular intervals. 

Author: No, polling is not enabled. Here, the idea is creating an HTTP step for the purpose of HTTP polling.

> HTTP step supports polling. When you create the HTTP step for polling, the client will keep requesting the resource at regular intervals. 

Author: ```suggestion
HTTP step supports polling. When you create the HTTP step for polling, the client requests the resource at regular intervals. 
```

To support HTTP polling: 

1. In the step's **Step Parameters >** **Optional Configuration**, enter the following details: 
    * **Assertion (optional)**: Enter the expression to validate the incoming response. You can use the following aliases to refer to the HTTP responses, URL, and method:
        * `<+httpsResponseCode>`
        * `<+httpUrl>`
        * `<+httpMethod>`
        * `<+httpResponseBody>`
    * **Headers (optional)**: Enter the key and value for the headers for the message. 
      

>       For example, in **Key**, enter token, in **Value**, enter secret references such as `<+secrets.getValue("aws-playground_AWS_secret_key")>`.

Author: If we say `token` that would mean we want the users to enter `token`. They're supposed to enter the token value. I'll change the text to .. in **Key**, enter the token...

    * **Output (optional)**: Create output steps to be used by other steps in the stage. 
      
      The **Value** setting can contain any HTTP step input, output, or response information. You can also use JSON and XML functors in the value.

> 2. In **Advanced >** **Failure Strategy**: 

Reviewer: ```suggestion
2. In **Advanced** > **Failure Strategy**, do the following: 
```
The UI element's name might have an arrow mark, but we should drop that from UI element names in writing. Then we can add the arrow to indicate a cascading menu. Looks odd otherwise. Also, check MMSTP at https://learn.microsoft.com/en-us/style-guide/procedures-instructions/formatting-text-in-instructions (says "If an option label ends with a colon or an ellipsis, don't include that end punctuation in instructions.")

>     * In **On failure of type**, select **All Errors**.

Reviewer: Doesn't the user have to enter something in the text box? Can they leave it blank and only select the check box?

Author: They can leave it blank if they're selecting All Errors checkbox.

    * In **Perform Action**, select **Retry**.

>     * Enter the **Retry count** and **Retry intervals**.

Reviewer: * Please let the UX team know that we should use the singular "intervals".
* Avoid using UI element names in flowing text. I don't know how else to describe it. Consider "In **Retry count** and **Retry intervals**, specify ..., respectively. (Or split the instruction into two instructions.)

    * In **Post retry failure action**, select **Ignore Failure**.
    
    ![](./static/http-polling.png)

> The HTTP step will retry polling and ignore all errors until the condition mentioned in the assertion section is met.

Reviewer: Avoid using the future tense as much as possible. The future tense is almost never required, and even then, there should be additional information about when that event will happen. 
```suggestion
The HTTP step retries polling and ignores all errors until the condition mentioned in the assertion section is met.
```

Capability checks are basic accessibility checks and do not follow multiple redirects. Hence, Harness returns from the first `302 Found` response during capability checks. 

## HTTP polling

The HTTP step supports polling. When you create the HTTP step for polling, the client requests the resource at regular intervals. 

To configure polling in the HTTP step: 

> 1. In the step, go to **Step Parameters** > **Optional Configuration**, and then enter the following details: 

Author: ```suggestion
1. In the step, go to **Step Parameters** > **Optional Configuration**, and enter the following details: 
```

    * **Assertion (optional)**: Enter the expression to validate the incoming response. You can use the following aliases to refer to the HTTP responses, URL, and method:
        * `<+httpsResponseCode>`
        * `<+httpUrl>`
        * `<+httpMethod>`
        * `<+httpResponseBody>`
    * **Headers (optional)**: Enter the key and value for the headers in the message. 
      
      For example, in **Key**, enter the token, in **Value**, enter secret references such as `<+secrets.getValue("aws-playground_AWS_secret_key")>`.

>     * **Output (optional)**: Create output steps to be used by other steps in the stage. 

Author: ```suggestion
    * **Output (optional)**: Create the output steps to be used by other steps in the stage. 
```
